rb_tree.py

```diff
@@ -1,81 +1,287 @@
-    NIL_LEAF = Node(value=None, color=NIL, parent=None)
+NIL_LEAF = Node(value=None, color=NODE_COLOR.NIL, parent=None)
 
-    def __init__(self):
+class RedBlackTree:
+    """
+    This is an implementation of the data structure described in the paper
+    Leonidas J. Guibas and Robert Sedgewick (1978). "A Dichromatic 
+    Framework for Balanced Trees". Proceedings of the 19th Annual 
+    Symposium on Foundations of Computer Science. pp. 8–21. 
+            doi:10.1109/SFCS.1978.3
+    """
+
+    def __init__(self) -> None:
         self.count = 0
         self.root = None
+        self.type = type(None)
+
+        # Used for deletion and uses the sibling's relationship with 
+        # his parent as a guide to the rotation
         self.ROTATIONS = {
-            # Used for deletion and uses the sibling's relationship with his parent as a guide to the rotation
             'L': self._right_rotation,
             'R': self._left_rotation
         }
 
-    def __iter__(self):
-        if not self.root:
-            return list()
+
+    def __iter__(self) -> Node:
+        """
+        Work our way through the Nodes by invoking
+        their __iter__ functions.
+        """ 
+
+        if not self.root: return []
         yield from self.root.__iter__()
 
-    def add(self, value):
+
+    def iter_dump(self) -> str:
+        if not self.root: return [""]
+        yield from self.root.iter_dump()
+
+
+    def __len__(self) -> int:
+        return self.count
+
+
+    def __stack_trace(self) -> str:
+        """ Easy to read, tabular output. """
+
+        exc_type, exc_value, exc_traceback = sys.exc_info()
+        this_trace = traceback.extract_tb(exc_traceback)
+        r = []
+        for i, _ in enumerate(this_trace):
+            r.append("""{}: {}, line {}, fcn {}\n    context=>> {}""".format(
+                i, _[0], _[1], _[2], _[3]
+                ))
+        return "\n".join(r)
+
+
+    def add(self, value:Any) -> None:
+        """
+        Add value to tree by creating a Node to contain it, and inserting
+        the node at the correct location in the tree.
+        """
         if not self.root:
-            self.root = Node(value, color=BLACK, parent=None, left=self.NIL_LEAF, right=self.NIL_LEAF)
+            """ The root Node is always BLACK """
+            self.root = Node.BlackNode(value)
             self.count += 1
+            self.value_type = type(value)
             return
+
+        if type(value) != self.value_type:
+            raise Exception(
+                "This tree contains {}. You cannot insert <{}> because it is a {}".format(
+                    self.value_type, value, type(value)
+                    ))
+
         parent, node_dir = self._find_parent(value)
-        if node_dir is None:
-            return  # value is in the tree
-        new_node = Node(value=value, color=RED, parent=parent, left=self.NIL_LEAF, right=self.NIL_LEAF)
+
+        if node_dir is None: # it is already here.
```

Comment:
nit: I prefer the previous comment `value is in the tree`. Maybe have it be `value is already in the tree` ?